### PR TITLE
Added support for master electron packages

### PR DESCRIPTION
**Pull request by bogthe**
Signed-off-by: Bogdan Stolojan <petre.stolojan@arm.com>

This PR fixes: https://github.com/theia-ide/theia/issues/3364
By adding a new extension method called `backendMasterElectron` to modules which will be required into node inside of the master electron process and enabling packages like `electron-updater` to work with Theia.

### CODE REVIEW

**Comment by thegecko** [comment]
Any feedback on this PR?

**Comment by bogthe** [comment]
Hi @kittaakos this seems to have been floating for a while now. Any feedback on having this merged?

File: dev-packages/application-package/src/extension-package.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ export interface Extension {
     frontendElectron?: string;
     backend?: string;
     backendElectron?: string;
+    backendMasterElectron?: string;
```
Comment (akosyakov) [inline dev-packages/application-package/src/extension-package.ts]: What are expectations about js fie, the same as for other entry points? Does it return `ContainerModule` and use inversify?

Comment (bogthe) [reply dev-packages/application-package/src/extension-package.ts]: As it stands it doesn't use `ContainerModule` or inversify.
Reason being I would like to see how the use cases for this entry point evolve and didn't want to impose any restrictions. For now each `backendMasterElectron` package can use it's own inversify containers but a simple js entry point would do for `electron-updates` (or any modules required to run from a different thread) to work.

**Comment by bogthe** [comment]
Hi @kittaakos , no worries! Glad I could help.

Check this branch out, I created a small [example here](https://github.com/ARMmbed/theia/compare/electron-master-extension...ARMmbed:electron-master-extension-example?expand=1).

File: dev-packages/application-manager/src/generator/backend-generator.ts [hunk 2]
```diff
@@ -72,6 +79,12 @@ module.exports = (port, host, argv) => Promise.resolve()${this.compileBackendMod
     });`;
     }
 
+    protected compileElectronMaster(backendMasterModules: Map<string, string>) {
+        return `// @ts-check
+${[...backendMasterModules.values()].map(module => `require('${module}')`)
```
Comment (kittaakos) [inline dev-packages/application-manager/src/generator/backend-generator.ts]:
```js
${[...backendMasterModules.values()].map(module => `require('${module}')`)
                .join('\n')}`;
```

should be

```js
${[...backendMasterModules.values()].map(module => `require('${module}');`).join('\n')}`;
```

(You have missed the semi-colons after the require.)

File: dev-packages/application-manager/src/generator/frontend-generator.ts [hunk 3]
```diff
@@ -223,7 +223,10 @@ if (isMaster) {
                 // See: https://github.com/theia-ide/theia/issues/835
                 process.kill(cp.pid);
             });
-        }
+        }${
+            this.pck.targetElectronMasterModules.size > 0 ?
+                `
+        require('../backend/electronMaster.js')` : ''}
```
Comment (kittaakos) [inline dev-packages/application-manager/src/generator/frontend-generator.ts]: Missing semi-colon after `require('../backend/electronMaster.js'')`.

File: dev-packages/application-manager/src/generator/backend-generator.ts [hunk 4]
```diff
@@ -22,6 +22,13 @@ export class BackendGenerator extends AbstractGenerator {
         const backendModules = this.pck.targetBackendModules;
         await this.write(this.pck.backend('server.js'), this.compileServer(backendModules));
         await this.write(this.pck.backend('main.js'), this.compileMain(backendModules));
+
+        if (this.pck.targetElectronMasterModules.size > 0) {
+            await this.write(
+                this.pck.backend('electronMaster.js'),
```
Comment (kittaakos) [inline dev-packages/application-manager/src/generator/backend-generator.ts]: Please use the [preferred file naming convention](https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names).

Also, why is it called electron master modules? I do not have a better idea, but `electronMaster` does not sound good to me.

Comment (bogthe) [reply dev-packages/application-manager/src/generator/backend-generator.ts]: Sorry 'bout that!

I have to say it doesn't sound that good to me either but I didn't know what else to name it. They are `packages` which will be ran from the master process of electron not the workers, and main process would've been confusing... so went with `master`! ¯\_(ツ)_/¯

**Comment by bogthe** [comment]
@kittaakos thanks for checking it! Added the changes you requested

File: dev-packages/application-manager/src/generator/backend-generator.ts (hunk 4 above)
Comment (kittaakos) [reply dev-packages/application-manager/src/generator/backend-generator.ts]: No, not the master process of electron, but the main process of electron. And later, the electron main process forks the backend process, and the forked backend process will create the workers...

Comment (kittaakos) [reply dev-packages/application-manager/src/generator/backend-generator.ts]: I would go with electron main (process) contributions... something like that. We can agree that the feature you want works 👍 but I would like to find a proper name for it before we merge it.

File: dev-packages/application-package/src/application-package.ts [hunk 5]
```diff
@@ -220,6 +228,10 @@ export class ApplicationPackage {
         return this.ifBrowser(this.backendModules, this.backendElectronModules);
     }
 
+    get targetElectronMasterModules(): Map<string, string> {
+        return this.ifElectron(this.backendElectronMasterModules, new Map<string, string>());
+    }
```
Comment (kittaakos) [inline dev-packages/application-package/src/application-package.ts]: This method isn't necessary here, right? I can understand the `targetFrontendModules` and the `targetBackendModules` but not this one. I think it is sufficient to have the other method (`backendElectronMasterModules`), and if not electron, then return with an empty map. You do not have electron master modules in the browser, so no need to mimic a symmetric API. What do you think?

**Comment by marechal-p** [comment]
Looking at this PR, it reminded me that I had been thinking about using inversify for the electron application in the same way that we do for the backend and frontend apps.

Do you think it would be worth to do sometime? Right now I really think it would be nice to have it modular like BE/FE, would allow clients have more control on the Electron feel without contributing to the generators everytime.

File: dev-packages/application-package/src/application-package.ts (hunk 5 above)
Comment (bogthe) [reply dev-packages/application-package/src/application-package.ts]: You're right, I'll get rid of the method.

**Comment by bogthe** [comment]
@marechal-p are you referring to the way BE and FE exchange messages through `JsonRpcConnectionHandler` and `WebSocketConnectionProvider`?
As it is you can still pass messages from any FE to a `backendMasterElectron` package through the basic `ipcRenderer / ipcMain`, which I admit is not ideal 😄 because it's different from the standard way. So yeah, would like to see this evolve into something like that, but for now it looks like a nice playground for things like update mechanisms, etc.

**Comment by bogthe** [comment]
woops

**Comment by marechal-p** [comment]
@bogthe I wasn't talking about the communication FE/BE, but rather the way they are constructed using inversify's container modules, which allows for contribution providers and symbol rebinding.

Currently the only way to extend the way the electron app part feels is by changing the generators, in the main repo. Another solution would be to forget about our generators and create your own. That could be different, that's my point.

File: dev-packages/application-package/src/extension-package.ts (hunk 1 above)
Comment (akosyakov) [reply dev-packages/application-package/src/extension-package.ts]: It's decided to Theia to use inversify in all entry points to allow customizations. I don't think this entry point should deviate.

Comment (thegecko) [reply dev-packages/application-package/src/extension-package.ts]: Can we use this entry point as a short term solution and replace it with an inversify method later?

**Pull request closed without merging**